### PR TITLE
starlark: disallow {+1} in format strings

### LINKED ISSUE #114: reject +0 as field identifier in string.format
```
$ starlark -c "'{+0}'.format(9)"
$ python3 -c "'{+0}'.format(9)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
KeyError: '+0'
$ python2 -c "'{+0}'.format(9)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
KeyError: '+0'
```

The spec says: `The *field name* may be either a decimal number or a keyword.` Decimal number isn't defined, although python rejects +0. starlark rejects other weird possible decimal numbers:

```
$ starlark -c "'{0.0}'.format(9)"
Traceback (most recent call last):
  cmdline:1: in <toplevel>
Error: attribute syntax x.y is not supported in replacement fields: 0.0
$ starlark -c "'{-0}'.format(9)"
Traceback (most recent call last):
  cmdline:1: in <toplevel>
Error: keyword -0 not found
```

So I think we should probably reject +0 (and +1 etc).

### PR DESCRIPTION
**Pull request by alandonovan**
+Test

Fixes #114

### CODE REVIEW

File: starlark/library.go [hunk 1]
```diff
@@ -1780,6 +1780,18 @@ func string_format(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Valu
...
 }
 
+func decimal(s string) (x int, ok bool) {
+	n := len(s)
+	for i := 0; i < n; i++ {
+		digit := s[i] - '0'
+		if digit > 9 {
+			return 0, false // not in [0-9]
+		}
+		x = x*10 + int(digit)
```
Comment (josharian) [inline starlark/library.go]:
```go
if x < 0 {
  return 0, false // overflow
}
```

File: starlark/library.go [hunk 2]
```diff
@@ -1780,6 +1780,18 @@ func string_format(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Valu
...
+	}
+	return x, 0 < n && n < 10
```
Comment (josharian) [inline starlark/library.go]:
```go
return x, true
```

File: starlark/testdata/string.star [hunk 3]
```diff
@@ -174,6 +174,8 @@ assert.eq("a{x}b{y}c{}".format(1, x=2, y=3), "a2b3c1")
 assert.fails(lambda: "a{z}b".format(x=1), "keyword z not found")
 assert.fails(lambda: "{-1}".format(1), "keyword -1 not found")
 assert.fails(lambda: "{-0}".format(1), "keyword -0 not found")
+assert.fails(lambda: "{+0}".format(1), "keyword \\+0 not found")
+assert.fails(lambda: "{+1}".format(1), "keyword \\+1 not found") # starlark-go/issues/114
```
Comment (josharian) [inline starlark/testdata/string.star]:
```python
assert.eq("{0000000000001}".format(0,1), "1")
```

(This matches Python 2/3 behavior, and makes explicit to future readers of the code that this behavior is intended.)

File: starlark/library.go (hunk 1 above)
Comment (alandonovan) [reply starlark/library.go]: The final condition (0 < n && n < 10) is false if there was any possibility of underflow.

Comment (alandonovan) [reply starlark/library.go]: Done.

File: starlark/library.go (hunk 2 above)
Comment (alandonovan) [reply starlark/library.go]: Done.

File: starlark/testdata/string.star (hunk 3 above)
Comment (alandonovan) [reply starlark/testdata/string.star]: Good point. This much better than a length check.

Comment (alandonovan) [reply starlark/testdata/string.star]: I added this one too:
assert.eq("{012}".format(*range(100)), "12") # decimal, despite leading zeros

**Pull request merged**